Target data config and schema proposal (#31)
reichlab/decisions: decisions/2025-06-17-RFC-target-data-metadata.md

```json
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "https://raw.githubusercontent.com/hubverse-org/schemas/main/v6.0.0/target-data-schema.json",
    "title": "Schema for Modeling Hub target data definitions",
    "description": "This is the schema of the target-data.json configuration file that defines metadata about target data used to visualise and evaluate modeling hub model outputs.",
    "type": "object",
    "description": "Target data metadata.",
    "properties": {
        "target_data_metadata": {
            "type": "object",
            "properties": {
                "time-series": {
                    "type": "object",
                    "properties": {
                        "observable_unit": {
                            "description": "Names of columns whose unique value combinations define an observable unit in time-series data. Each combination of values must be unique across `as_of` data versions. The majority are expected to correspond to task ID names but may include other columns as well (e.g. the general `date` column).",

>                             "type": "array",

Reviewer: I understand the description, but just a small notice, not all the time series will have `as_of` column amd maybe to be clearer we can add:
```suggestion
                            "description": "Names of columns whose unique value combinations define an observable unit in time-series data. Each combination of values must be unique. If multiple values available for the same time point they should be unique across `as_of` data versions. The majority are expected to correspond to task ID names but may include other columns as well (e.g. the general `date` column).",
                            "type": "array",
```
If you think it's more confusing, feel free to ignore

                    "required": ["observable_unit"],
                    "additionalProperties": false
                }
            },
            "required": ["time-series", "oracle-output"],
            "additionalProperties": false
        }
    }
}
```

To summarise, the `target-data.json` file will contain a `target_data_metadata` object with two main properties: `time-series` and `oracle-output`. Each property will define metadata about the respective target data type.

- **`time-series`**: This object contains three main main properties:
    - `observable_unit`: An array of columns whose unique value combinations define an observable unit in time-series data. Each combination of values must be unique across `as_of` data versions. In time-series data, this typically includes task IDs that uniquely identify a specific observation at a given point in time but might include columns that do not directly correspond to task ID names (e.g. a more general `date` column).
    - `extra_task_ids`: An optional array of task IDs that are not part of the observable unit but are present in the time-series data. These task IDs may be used for additional context or filtering.

>     - `non_task_id_schema`: An object that defines the schema for non-task ID columns in the time-series data. Each property in this object represents a non-task ID column name, and its value is the data type of that column. The data types represent R data types can be one of the following: `character`, `double`, `integer`, `logical`, or `Date`. This will allow for the inclusion of additional columns in the time-series data that do not correspond directly to task IDs. Note that the standard column `observation` DOES NOT need defining/including as will always be cast as `double`. It is also never part of the observable unit.

Reviewer: Just by curiosity, are we automatically casting `"observation"` as `double`? because it some cases we might have `integer` (death count for example).

Author: `double` is just safer and means we don't have to define this column also. R can always do maths and even joins with `integer` and `double` data but we never need to merge on `observation` in time-series and `value` in model output data anyways.

• checks: FALSE
• file_system: "S3FileSystem"
• model_output_dir: "model-output/"
• config_admin: hub-config/admin.json
• config_tasks: hub-config/tasks.json

── Connection schema 
hub_connection with 2454 Parquet files
9 columns
reference_date: date32[day]
target: string
horizon: int32
location: string
target_end_date: date32[day]
output_type: string
output_type_id: string
value: double
model_id: string
```

##### `time-series`

```
target_timeseries with 1 csv file
7 columns
as_of: date32[day]
target: string
target_end_date: date32[day]
location: string
location_name: string
observation: double
weekly_rate: double
```
##### `oracle-output`

```
target_oracle_output with 1 csv file
8 columns
as_of: date32[day]
target: string
target_end_date: date32[day]
location: string
horizon: int32
output_type: string
output_type_id: string
oracle_value: double
```

The proposed example [`target-data.json`](./2025-06-17-RFC-target-data-metadata/flusight-target-data.json) file for the Flusight hub would look like this:

```json
{
    "target_data_metadata": {
        "time-series": {
            "observable_unit": [
                "target",
                "target_end_date",

>                 "location"

Reviewer: Following the documentation, I don't understand why the column `"as_of"`, `"location_name"` are not included in the `"observable_unit"`.  I understand that `"target"`, `"target_end_date"`, `"location"` are the minimal observable unit, but why don't you want to add location name to be validated with the same logic as location? and `as_of` also defined an observable unit for me as if present it's what's used to define unique observable unit?

Author: Let's take `location_name` first. Because `location_name` is an additional column that is not part of `task_ids` (representing the official definition of tasks) and is also directly related to the `location` code (so a bit of a derived variable). So it doesn't add anything to the observable unit as it relates to the model output data.  The `as_of` column again is more of a versioning variable that doesn't directly relate to forecasting observations in my view. But given we know the standard data type for it, I would be open to including it if others thought it makes sense too

Reviewer: I agree with your comment, what I am wondering here is for the validation. The JSON file here will be used in the validation too, if I understand correctly. 
As the `as_of` is optional and impact the observable_unit, it makes sense to me to have that information at least somewhere particular in the JSON, whether it's in the `observable_unit` or somewhere else. I think it's important to have it noted somewhere.
For the `location_name`, I understand it as a kind of derived variable. However, if a team want it to be validated with the `observable_unit` to check that they are "unique" when necessary, it will make sense to me to put it in the `observable_unit`. I guess it can be at team discretion to decide what to do about the "derived" column.

Reviewer 2: I agree with @annakrystalli here. I think we want the observable-unit property to represent the minimal observable unit.

Author: For me the important thing is that they do not form part of the modeling task as defined in the config and therefore to not relate to merging data with model output data and should therefore not be part of the observable unit. I think this can be a helpful heuristic for hub admins to use to determine what should and should not be part of the observable unit

# 2025-06-17 RFC Time-series target data schema metadata

## Context

> While moving forward with validation tests for target data, we have identified the need for a structured metadata schema that can be used to validate time-series target data schema deterministically. In particular, our allowance of inclusion of non-task ID columns in time-series data has so far required the opening of target time-series data to extract the default schema for such non-task ID columns when creating a time-series schema. This has uncovered a number of challenges/limitations:

Reviewer 2: Would it be fair to say that part of the need for this is due to inclusion of non-task-id columns, but part of it is also due to hive partitioning on, say, a date? if so, maybe we could change the text here to reflect that.

Author: I think that's explicitly covered in point 2 below no?

Reviewer 2: Yes, but I was mostly reacting to the way it is phrased in this sentence where it sounds like it is just due to the non-task-id columns issue and doesn't mention the partioning:
> In particular, our allowance of inclusion of non-task ID columns in time-series data has so far required the opening of target time-series data to extract the default schema for such non-task ID columns when creating a time-series schema.

Author: The thing is, the main issue with the partitioning arises from the fact we need to open the dataset without a schema (with default schema settings - all characters - to extract the default data types. If we weren't opening the dataset, it wouldn't be a problem. it's also a special case, where data were being partitioned on columns that exist in the data and are not character. This in combination with the attempt to open the dataset with default settings is what creates the problem and causes it to fail. So I still feel it's better covered in the more detailed point 2:
> 2. Problems arising from hive partitioning on columns that also are contained in the data e.g. the nowcast hub which is partitioned on a date type column. The parquet files contain the correct data type (date) for that column but the default dataset behaviour when opening datasets without an explicit schema is to cast all partition variables as character. This causes a conflict during the attempt to open the dataset resulting in the dataset not being able to be opened without an explicit schema defined. Because of this, such datasets cannot currently be opened with our current hubData functions https://github.com/hubverse-org/hubData/issues/89

But I'm happy with a suggestion though of what it should say?

Author: Just saw your suggestion and committed. Thanks!

- listing any columns in the target data type that correspond to task ID variables (whose data types are already defined via `tasks.json`)
- listing any non-task ID columns along with their data types that are present in the time-series data. 

The [**target-data-schema.json**](./2025-06-17-RFC-target-data-metadata/target-data-schema.json) file describing the properties of the a `target-data.json` config file will be as follows:

```json
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "https://raw.githubusercontent.com/hubverse-org/schemas/main/v6.0.0/target-data-schema.json",
    "title": "Schema for Modeling Hub target data definitions",
    "description": "This is the schema of the target-data.json configuration file that defines metadata about target data used to visualise and evaluate modeling hub model outputs.",
    "type": "object",
    "description": "Target data metadata.",
    "properties": {
        "target_data_metadata": {
            "type": "object",
            "properties": {
                "time-series": {
                    "type": "object",
                    "properties": {
                        "observable_unit": {

>                             "description": "Names of columns whose unique value combinations define an observable unit in time-series data. Each combination of values must be unique. If multiple values available for the same time point they should be unique across `as_of` data versions. The majority are expected to correspond to task ID names but may include other columns as well (e.g. the general `date` column).",

Reviewer 2: Could we reference a point in the documentation where we define what a observable unit is?

Author: Do you mean something like the here? https://docs.hubverse.io/en/latest/user-guide/target-data.html#time-series

It feels like mabe a box defining it with more detail might be beneficial in our docs?

I also notice we're calling it unit of observation in our docs. Do we have a preference on the term as it would be best to use the same terminology throughout?

Author: I've added this as a task to the following issue: https://github.com/hubverse-org/hubDocs/issues/382

>                             "description": "Names of columns whose unique value combinations define an observable unit in time-series data. Each combination of values must be unique. If multiple values available for the same time point they should be unique across `as_of` data versions. The majority are expected to correspond to task ID names but may include other columns as well (e.g. the general `date` column).",

Reviewer 2: ```suggestion
                            "description": "Names of columns whose unique value combinations define an observable unit in time-series data. Each combination of values must be unique. If multiple values are available for the same time point they should be unique across `as_of` data versions. The majority are expected to correspond to task ID names but may include other columns as well (e.g. the general `date` column).",
```

                            "type": "array",
                            "uniqueItems": true,
                            "items": {
                                "type": "string"
                            }
                        },
                        "extra_task_ids": {
                            "description": "Names of task IDs that are not part of the observable unit but are present in the time-series data. These task IDs may be used for additional context or filtering.",
                            "type": ["array", "null"],
                            "uniqueItems": true,
                            "items": {
                                "type": "string"
                            }
                        },
                        "non_task_id_schema": {
                            "type": "object",
                            "uniqueItems": true,
                            "additionalProperties": {
                                "type": "string",
                                "enum": ["character", "double", "integer","logical", "Date"],

>                                 "description": "Non-task ID column names and data types found in time-series data. Include any columns in the time-series data that does not correspond exactly to a task ID. If an `as_of` column is included, it should be specified here as well."

Reviewer 2: I can't understand from this description how the content of this field should be formatted. Is it asking for pairs of column names and data types?

Author: Fair point. I've added a clarifying comment and a couple of examples in 6793cca86439f457163e7fccd646617a109468c0

                    "additionalProperties": false
                },
                "oracle-output": {
                    "type": "object",
                    "properties": {
                        "observable_unit": {
                            "description": "Names of task IDs whose unique value combinations define an observable unit in oracle output data. Each combination of values must be unique once combined with output type IDs.",
                            "type": "array",
                            "uniqueItems": true,
                            "items": {
                                "type": "string"
                            }
                        },
                        "extra_task_ids": {
                            "description": "Names of task IDs that are not part of the observable unit but are present in the oracle-output data. These task IDs may be used for additional context or filtering (e.g. some derived task IDs).",
                            "type": ["array", "null"],
                            "uniqueItems": true,
                            "items": {
                                "type": "string"
                            }

>                         },

Reviewer 2: Won't the observable unit and extra task ids be the same across both time-series and oracle-output data? If so, maybe those could be moved to a top-level category so the content doesn't have to be repeated. In both examples I think they are the same (or should be - in the flusight example there is a "horizon" variable included but I think it can/should be left out).

Author: I considered that initially but it turns out they are not necessarily always the same. That's because we allow a non task ID column in time-series data e.g [a column called `date`](https://docs.hubverse.io/en/latest/user-guide/target-data.html#time-series).  Hubs don't always have a `target_end_date/target_date` column to match in the `time-series` observable unit (e.g. they might only have `origin_date` + `horizon` in their task IDs and model outputs so in that case the `time-series` data might just have a `date` column.

Having said that we could allow either a top-level property when they are the same and a target data type property for when they differ?

Reviewer 2: Yes, I agree with your second suggestion. So basically they are always the same task-ids but in time-series we allow for a column representing a date but with a different name from the task-id column to stand in for a task-id column. E.g.

The `target_end_date` column is part of the observable unit, but in the time-series target data we allow for `target_end_date` to be replaced by `date`.

I think that these are separate parts of the metadata. That is to say that:

1. the observable unit columns should be defined once in a top-level property
2. if the date column is renamed in time-series file that should be specified in a "time-series-metadata" subsection of the target-data config file.

Author: I like the idea of handling the date column separately. I think we should build up some documentation regarding this special column and:
    1. how it relates `time-series` data to `oracle-output` data
    2. how it relates `model-output` data to both `time-series` data to `oracle-output` data.
    3. There's still to think through the situation where a hub is only collecting `origin_date` and `horizon`. Model outputs will need to have a column of actual `target_date`s to merge to target data. We should probably document this somewhere. 
   
Should I try and capture this in this document first or open issues?

                    },
                    "required": ["observable_unit"],
                    "additionalProperties": false
                }
            },
            "required": ["time-series", "oracle-output"],
            "additionalProperties": false
        }
    }
}
```

To summarise, the `target-data.json` file will contain a `target_data_metadata` object with two main properties: `time-series` and `oracle-output`. Each property will define metadata about the respective target data type.

- **`time-series`**: This object contains three main main properties:

>     - `observable_unit`: An array of columns whose unique value combinations define an observable unit in time-series data. Each combination of values must be unique across `as_of` data versions. In time-series data, this typically includes task IDs that uniquely identify a specific observation at a given point in time but might include columns that do not directly correspond to task ID names (e.g. a more general `date` column).

Reviewer: ```suggestion
    - `observable_unit`: An array of columns whose unique value combinations define an observable unit in time-series data. Each combination of values must be unique. If multiple values available for the same time point they should be unique across `as_of` data versions. In time-series data, this typically includes task IDs that uniquely identify a specific observation at a given point in time but might include columns that do not directly correspond to task ID names (e.g. a more general `date` column).
```

Reviewer: to match the description and clarify that `as_of` is optional

output_type_id: string
value: double
model_id: string
```

##### `time-series`

```
target_timeseries with 1 csv file
7 columns
as_of: date32[day]
target: string
target_end_date: date32[day]
location: string
location_name: string
observation: double
weekly_rate: double
```
##### `oracle-output`

```
target_oracle_output with 1 csv file
8 columns
as_of: date32[day]
target: string
target_end_date: date32[day]
location: string
horizon: int32
output_type: string
output_type_id: string
oracle_value: double
```

The proposed example [`target-data.json`](./2025-06-17-RFC-target-data-metadata/flusight-target-data.json) file for the Flusight hub would look like this:

```json
{
    "target_data_metadata": {
        "time-series": {
            "observable_unit": [
                "target",
                "target_end_date",
                "location"
            ],
            "non_task_id_schema": {
                "as_of": "Date",
                "location_name": "character",
                "weekly_rate": "double"
            }
        },
        "oracle-output": {
            "observable_unit": [
                "target",
                "target_end_date",
                "location",

>                 "horizon"

Reviewer 2: I don't think horizon is needed here.

Author: I think it is because it exists in the `oracle-output` data. If we didn't include it, we would not be able to validate the data for duplicates. You could argue that it's not required in the `oracle-output` data but given that it is I feel it needs to be included in the observable unit

Reviewer 2: I guess this is related to the question above about whether the task-id columns can be different for oracle and time-series data.

I think horizon is not and should not be included in oracle output data. Are you saying that horizon is included in the actual oracle output data that is being stored in the flusight hub currently? I don't think we should compromise on what our schema __should__ be because a hub isn't doing it correctly right now.

Author: Sorry yes, `horizon` is indeed part of the `oracle-output` data:
```
target_oracle_output with 1 csv file
8 columns
as_of: date32[day]
target: string
target_end_date: date32[day]
location: string
horizon: int32
output_type: string
output_type_id: string
oracle_value: double
```

I'm certainly in favour of enforcing the minimum required information to merge to `model-ouput` data/`oracle-output` data.

# 2025-06-17 RFC Time-series target data schema metadata

## Context

> While moving forward with validation tests for target data, we have identified the need for a structured metadata schema that can be used to validate time-series target data schema deterministically. In particular, our allowance of inclusion of non-task ID columns in time-series data has so far required the opening of target time-series data to extract the default schema for such non-task ID columns when creating a time-series schema. This has uncovered a number of challenges/limitations:

Reviewer 2: ```suggestion
While moving forward with validation tests for target data, we have identified the need for a structured metadata schema that can be used to validate time-series target data schema deterministically. In particular, a combination of our allowance of inclusion of non-task ID columns in time-series data and supporting of hive-partitioned datasets has so far required the opening of target time-series data to extract the default schema for such non-task ID columns when creating a time-series schema. This has uncovered a number of challenges/limitations:
```

    "$id": "https://raw.githubusercontent.com/hubverse-org/schemas/main/v6.0.0/target-data-schema.json",
    "title": "Schema for Modeling Hub target data definitions",
    "description": "This is the schema of the target-data.json configuration file that defines metadata about target data used to visualise and evaluate modeling hub model outputs.",
    "type": "object",
    "description": "Target data metadata.",
    "properties": {
        "target_data_metadata": {
            "type": "object",
            "properties": {
                "observable_unit": {
                            "description": "Names of columns whose unique value combinations define the minimum observable unit in time-series data. Each combination of values must be unique across `as_of` data versions if applicable. The majority are expected to correspond to task ID names but may include other columns as well (e.g. the general `date` column).",
                        "type": "array",
                        "uniqueItems": true,
                        "items": {
                            "type": "string"
                        }
                    },
                "date_col": {

>                     "description": "Default name of the date column across hub data (time-series, oracle-output and model output). This is the column that stores the date on which observed data actually occured.",

Reviewer 3: Is the date column supposed to have a unified name across time series, oracle output, and model output? Or it is the date column separate for time series and model output (with the flexibility to be called just "date" in time series)?

Author: I think ideally yes, they are the same across all target data and model output data so then you can set it as the default. But we offer an option to override it in individual target types. With model outputs it's a bit trickier because it's possible that a hub would only store `reference_date` and `horizon` and not `target_date` (the default). A `target_date` column would need to be computed in the model output data too to be able to match them to target data. 

What are the expectations of the `hubEvals` functions for model outputs and oracle outputs?

Reviewer 3: From what I can tell, this is handled in scoringutils and it seems to just expect that the name of the date column in the oracle output matches that in the model output

Author: It might be useful to add a little note in our  hubEvals docs then to explain that expectation.

Reviewer 2: I think there have been conversations in the past that have suggested that the flexibility to have the `time-series` target data file data column be named something more generic like `date` rather than the rather hubverse-specific `target_end_date`. This generic name might feel more natural to someone trying to use the file as a modeling input, say. 

E.g., in the hubVis package, we support the `x_target_col_name` argument which is designed to specify what the date column is (and I think implicitly assumes it might be different from the `x_col_name` from the model output data.

From the perspective of enforcing a single, simple standard, I think just having the tools expect/require that there is a column defined as the "date corresponding to the observation" in model_output data and target data is a reasonable thing.

FWIW, it happened in my setting up of the recent sandbox hub that we did not have a target_end_date in the first pass of the data. I had to add it to use the predtimechart viz (which requires it). 

Given that it seems we are baking in this assumption in multiple places, I think it is fair to add this as an explicit requirement, even if it might be a little bit "awkward" in the time-series data file. 

So then, I make this suggestion, which just removes the word "default" since we are just saying this IS the name of the date field throughout all data, correct? 

```suggestion
                    "description": "Name of the date column across hub data (time-series, oracle-output and model output). This is the column that stores the date on which observed data actually occured.",
```

Reviewer 2: Also, we would require that `date_col` is present in the `observable_unit` values, right?

                    "type": ["string", "null"],
                    "default": null
                },
                "versioned": {

>                     "description": "Indicates whether target data are versioned using `as_of` dates. If true, the data is expected to have a date `as_of` column that indicates the version of each data point.",

Reviewer 3: Does this versioned property apply to both time series and oracle output formats?

Author: Yes. I think it's just easier to enforce like this.

>                     "description": "Indicates whether target data are versioned using `as_of` dates. If true, the data is expected to have a date `as_of` column that indicates the version of each data point.",

Author: ```suggestion
                    "description": "Indicates whether target data are versioned using `as_of` dates. If true, the both time-series and oracle-output data are expected to have a date `as_of` column that indicates the version of each data point.",
```

Is this clearer maybe @lshandross ?

Reviewer 3: Yes, thank you!

                    "type": "boolean",
                    "default": false
                },
                "time-series": {
                    "type": "object",
                    "properties": {

>                         "extra_task_ids": {

Reviewer 2: Could we give an example here that corresponds to the examples given in the non_task_id_schema below?

Author: Added in a39a5eff91f1dba19df57880ba22a5d7a497a29d. Having said that I'm not sure it's a good example as it's not something that would be included in time-series. In all honesty am having difficulty thinking of any task IDs that are not part of the observable unit but would make sense to be included in the time-series data. Any examples you can think of? Perhaps this property isn't necessary at all?

Reviewer 3: I can't think of anything, but maybe Lucie has a use case for SMH that we want to support?

Author: @LucieContamin any thoughts on the above discussion?

Reviewer: The only thing I can think of is redundant columns, but I cannot think of a good example. 

I guess you can add an `age_group` column that cover only one age group (for example the overall population), if you have a round with age group but don't have detailed target data. It's not exactly in the definition, it will still be part of the observable unit, it just define the overall population in age unit, so I don't know.

Author: Yeah I thought about that situation but in the case where target data cannot be matched to age groups, then an age group column isn't adding any useful information so doesn't feel like it should be in the time-series data. On the other hand if it does have  useful age group data to merge onto model outputs than it should be part of the observable unit. Overall the examples we've come up with seem contrived and might prove more confusing and require a lot of explaining to be clear so feel like they would have the opposite effect of what their designed for, i.e. to demonstrate clearly how the property is useful.

It still makes me feel this is not a useful property and should be removed. @nickreich what are your thoughts?

Reviewer 2: I agree that removing this property seems reasonable for now.

Author: Removed in a9633a5085fa742dad04864e97f3d42f8be56c94

                            "description": "Names of task IDs that are not part of the observable unit but are present in the time-series data. These task IDs may be used for additional context or filtering.",
                            "type": ["array", "null"],
                            "uniqueItems": true,
                            "items": {
                                "type": "string"
                            }
                        },
                        "non_task_id_schema": {
                            "type": "object",
                            "uniqueItems": true,
                            "examples": [
                                {
                                    "location_name": "character"
                                },{
                                    "date": "Date"
                                }
                            ],

>                             "additionalProperties": {

Reviewer 2: I don't understand schemas enough to see why this has to be under "additionalProperties" and why the description can't just be part of the "non_task_id_schema" itself.

Author: Because the names of the properties (i.e. the columns we are specifying data types for) are not known a priori or standard so we can't hardcode their expectation as we do for example for standard task IDs. This is the same approach we use to define expectations for custom task IDs.

Author: I moved the description though to the `non_task_id_schema ` property itself i 8b68a1f4526b46a17036007786129f205e713a58

                                "type": "string",
                                "enum": ["character", "double", "integer","logical", "Date"],
                                "description": "Key-value pairs of non-task ID column names and data types found in time-series data. Include any columns in the time-series data that does not correspond exactly to a task ID. If an `as_of` column is included, it should be specified here as well."
                            }
                        },

>                         "date_col": {

Reviewer 2: is this still needed here, or given the discussion above, could it just live at the top level?

Author: Removed!

                            "description": "Default name of the date column in time-series data. This is the column that stores the date on which observed data actually occured.",
                            "type": ["string", "null"],
                            "default": null
                        },
                    },
                    "additionalProperties": false
                },
                "oracle-output": {
                    "type": "object",
                    "properties": {

>                         "output_type": {

Reviewer 3: I don't love "output type" as the name for this property because it doesn't feel intuitive as to what it actually is meant to do. Can we change it to something like "includes_output_type_id"?

Author: I guess I went for the higher level property (`output_type`) in the config  because I felt it better encompasses both columns (`output_type` and `output_type_id`) that are both required. Were effectively saying output type specific information is required in our oracle output data?

Reviewer 3: How about something like "output_type_info"? As is, I feel like the intuitive read is that the property lists which output types the oracle output contains

Author: I think I still prefer the simpler `output_type`, plus we'll have documentation and explain what that setting is too. 

Maybe we could ask for a tie breaker from @nickreich ?

Author: How about `has_output_type_ids` ? Given it's a boolean seems more self explanatory but still short name? And thinking it through you're right, that it's the output type IDs that are of importance here.

Reviewer 2: I like `has_output_type_ids`, but do we need this field at all? Could the expected presence/absence not just be inferred from the types of data in the hub?

Reviewer 3: I'm good with either `has_output_type_ids` or inferring, if that's possible

Author: Inferring is not currently bullet proof because we are not being strict about not including an `output_type_id` column with `NAs` if only e.g. quantiles, means, medians and samples are collected. We can only say that they `must` be present if `cdf` or `pmf` output types are collected but that's only if a hub actually wants to score such targets. For now I've changed the property to `has_output_type_ids` but we could infer the requirement for these columns if we were strict about not including an `output_type` and `output_type_id` column if not necessary.

                            "default": false},
                        "observable_unit": {
                            "description": "Names of task IDs whose unique value combinations define an observable unit in oracle output data. Each combination of values must be unique once combined with output type IDs. Can be used to override default observable units in situations where some output types require additional task ID value to map onto target data.",
                            "type": "array",
                            "uniqueItems": true,
                            "items": {
                                "type": "string"
                            }
                        },
                        "date_col": {
                            "description": "Default name of the date column in oracle-output data. This is the column that stores the date on which observed data actually occured.",
                            "type": ["string", "null"],
                            "default": null

>                         },

Reviewer 2: We can delete "observable_unit" and "date_col" from here, right? because they are represented at the top level?

                    },
                    "required": ["output_type"],
                    "additionalProperties": false
                }
            },
            "required": ["observable_unit", "date_col"],
            "additionalProperties": false
        }
    }
}
```

> ## Summary of `target-data.json` Structure

Reviewer 2: Noting that I would suggest moving this to ABOVE the example. Provide the summary of what to expect, then details.

Author: Done!

The `target-data.json` file defines a `target_data_metadata` object with top-level properties that describe expectations across target datasets.

### Top-Level Properties

* `observable_unit`: An array of column names whose unique value combinations define the minimum observable unit. Must be task IDs or defined in the `date_col` property. Combinations must include the `date_col` (and `as_of` if versioned). `as_of` is never included in the observable unit as it is not a task ID but a versioning column. This property is required.

> * `date_col`: The default date column across time-series, oracle-output, and model-output datasets. Expected to be of type `Date`.

Reviewer 3: See my above comment on the date column

* `versioned`: Boolean indicating whether `as_of` versioning is used. If true, datasets must have a date `as_of` column indicating the version of each data point. Defaults to `false`. 

> I also proposing to allow an `as_of` column in `oracle-output` to support traceability if versioning is being used. it will allow us to link individual oracle value observations to the specific version of time-series data it was derived from. I propose we enforce that there should only be a single version of an observation in oracle output data so no filtering on `as_of` date is required to get a single version of available data.

Reviewer 3: I think I'm ok with this decision, but let's investigate about whether it will cause any current functionality to break similar to [what happened here](https://github.com/hubverse-org/hubPredEvalsData-docker/pull/5) (I'm thinking about hubEvals and dashboard stuff mostly)

Author: Fair point. Looking into the resolution of that PR, if there is an `as_of` column in ~~model~~ oracle output data, it's removed before creating to processing oracle output data so should affect that issue. It does seem like most hubs are already using it and it makes sense for traceability. But I agree we should investigate implications down stream before releasing the schema and fixing them. Could be a good opportunity to start working with the dashboard code too.

Reviewer 2: I think you mean "...if there is an `as_of` column in ORACLE OUTPUT data..." (not "model output").

I think it's ok to assume that most hubs will not use as_of in oracle output.

Author: In the proposal we're actually requiring that if the target data is versioned then both `time-series` and `oracle-output` have an `as_of` column but with oracle data containing only a single version of each observation while `time-series` can have multiple versions. This enables traceability of data points in oracle outputs back to the timeseries data they were derived from. It also makes it simpler. One setting `versioned: true` makes the expectation clear in both datasets

Reviewer 2: I don't think we want to ever require oracle-output data to be versioned or have as-of. It is a fairly fringe use-case when oracle output data need to be versioned, and we've made explicit design decisions elsewhere to not encourage use of this.

Author: Resolved in the hubverse dev meeting 2025-09-25. Proposal updated in 2b3ad7d26a7f8db7ae902dd59cb96e8005ad4903

### Target-Type Specific Configuration

* **`time-series`**:

>   * `date_col`: Optional override of the top-level `date_col`.

Reviewer 3: If we're allowing the option to override the top level date column property for both time series and oracle output, why are we even including the top level one? It feels just easier to have the two separate ones

Author: Because if they are the same for both, you only have to set it once at the top level.

Reviewer 3: I guess I can see the logic of this, but some part of me still doesn't love it. Not a blocker though

Reviewer 2: I think we should just have it at the top level and enforce that it is the same everywhere.

Reviewer 2: I could maybe be talked into allowing it to be different for time-series target data, but given that we're already requiring it to be the same for predtimechart, maybe just better to enforce it to be the same for simplicity.

Reviewer 2: actually, I see now and understand how within this one hub the observeable unit is different for different targets. yeesh.

Author: I think these are two different things. Indeed in rare cases (as is the horizon being a functional part of how the pmf output type is calculated) different observable units are required per target type. But that doesn't stop up from requiring that the date columns share the same name. 

The only slightly annoying implication of enforcing this is that the example complex forecasting hub which a lot of functionality testing is based on uses `date` in the time-series target data so many tests will need to be updated. Better to decide this now though than later. There may also be implications for some hubs/dashboards which we'll need to investigate.

Personally I also vote for keeping this consistent across target type datasets

Reviewer 3: I agree with enforcing consistency across target data

Author: Sure. I've removed `date_col` everywhere apart from the top level

  * `extra_task_ids`: Additional task IDs used for filtering or grouping.
  * `non_task_id_schema`: key-value pairs of non-task id column names and their R-data types, one of (`character`, `double`, `integer`, `logical`, `Date`). The `as_of` column does not need defining here.

* **`oracle-output`**:

  * `output_type`: Boolean. Must be true if `pmf` or `cdf` output types exist. Can be false otherwise. If true, the dataset must include `output_type` and `output_type_id` columns. Defaults to `false`.
  * `observable_unit`: Task IDs whose combination plus output type IDs, if applicable, uniquely define a row. This can be especially useful for [output types whose values are functionally dependent on other task IDs](https://github.com/reichlab/flusight-dashboard/issues/20#issuecomment-2815550603. 

>   * `date_col`: Optional override of the top-level `date_col`.

Reviewer 2: I think these could be removed from the oracle-output specific section.

Reviewer 2: correction: I think the `date_col` could still be removed here and required to be the same, but I guess I see the logic for the `observable_unit` to be different.

## Validations

> In addition to JSON Schema validation, the following dynamic checks will be applied:

Reviewer 2: add that `date_col` must be contained within `observeable_unit`.

Author: Done

> * `observable_unit` must only include task ID columns and `target_col`.

Reviewer 2: what is `target_col`?

Author: It's the column that contains details of the targets as specified in `target_keys`.

Author: I've updated the text to clarify that this is only required if the hub contains a `target` column. In situations where `target_keys: null` in target metadata, which implies a single global target for the whole hub which is not contained in the model output, there is no `target_col` to be included so it's not required.

* List columns that correspond to task IDs (whose data types are already defined via `tasks.json`);
* List non–task ID columns along with their data types.

## Summary of `target-data.json` Structure

The `target-data.json` file defines a `target_data_metadata` object with top-level properties that describe expectations across target datasets.

### Top-Level Properties

* `observable_unit`: An array of column names whose unique value combinations define the minimum observable unit. Must be task IDs and additionally always include the column defined in the `date_col` property. Unique combinations. If versioned, unique combinations will also take into account the values in the `as_of` column but is never included in the observable unit as it is not a task ID but a versioning column. This property is required.
* `date_col`: The default date column across time-series, oracle-output, and model-output datasets. Expected to be of type `Date`.
* `versioned`: Boolean indicating whether `as_of` versioning is used. If true, datasets must have a date `as_of` column indicating the version of each data point. Defaults to `false`. 

> I also proposing to allow an `as_of` column in `oracle-output` to support traceability if versioning is being used. it will allow us to link individual oracle value observations to the specific version of time-series data it was derived from. I propose we enforce that there should only be a single version of an observation in oracle output data so no filtering on `as_of` date is required to get a single version of available data.

Reviewer 3: ```suggestion
I am also proposing to allow an `as_of` column in `oracle-output` to support traceability if versioning is being used. it will allow us to link individual oracle value observations to the specific version of time-series data it was derived from. I propose we enforce that there should only be a single version of an observation in oracle output data so no filtering on `as_of` date is required to get a single version of available data.
```

### Target-Type Specific Configuration

* **`time-series`**:

>   * `non_task_id_schema`: key-value pairs of non-task id column names and their R-data types, one of (`character`, `double`, `integer`, `logical`, `Date`). The `as_of` column does not need defining here as it is expected to always be a date column.

Reviewer 2: ```suggestion
  * `non_task_id_schema`: key-value pairs of non-task id column names that will be present in the time-series target data dataset and their R-data types, one of (`character`, `double`, `integer`, `logical`, `Date`). The `as_of` column does not need defining here as it is expected to always be a date column.
```

  * `versioned`: Boolean indicating whether `as_of` versioning is used. If true, time-series data must have a date `as_of` column indicating the version of each data point. Defaults to `false`. 

* **`oracle-output`**:

  * `has_output_type_ids`: Boolean. Must be true if `pmf` or `cdf` output types exist. Can be false otherwise. If true, the dataset must include `output_type` and `output_type_id` columns. Defaults to `false`.
  * `observable_unit`: Task IDs whose combination plus any output type IDs if present, uniquely define a row. This can be especially useful for [output types whose values are functionally dependent on other task IDs](https://github.com/reichlab/flusight-dashboard/issues/20#issuecomment-2815550603.

>   * `versioned`: Boolean indicating whether `as_of` versioning is used. If true, oracle-output data must have a date `as_of` column indicating the version of each data point. Defaults to `false`. Note that atleast for now, oracle output data is expected to have only a single version of each unique combination of observable unit values in contrast to time-series which is allowed to have multiple. 

Reviewer 2: ```suggestion
  * `versioned`: Boolean indicating whether `as_of` versioning is used. If true, oracle-output data must have a date `as_of` column indicating the version of each data point. Defaults to `false`. Note that atleast for now, oracle output data is expected to have only a single version of each unique combination of observable unit values in contrast to time-series which is allowed to have multiple. This is to minimize confusion and reduce the risk of downloading multiple observed values and scoring on each of them.
```

        }
    }
}

```

## Validations

In addition to JSON Schema validation, the following dynamic checks will be applied:

> * `observable_unit` must only include task ID columns and the `date_col` and the `target_col` unless `target_keys` are `NULL` which implies a single global target and no `target` column.

Reviewer 3: I'd find this description to be a bit more understandable if we set something like "must only include the task ID columns, including the specified `date_col` and `target_col`, unless..." since the date and target columns specified in this schema are task ids

Author: So the reason I've worded it as such is because it's not guaranteed that `date_col` be a task ID. For example, a hub could only be collecting `origin_date` and `reference_date` in model outputs. In such a case, `target_end_date` would need to be calculated from `origin_date` and `reference_date` before plotting and evaluating but we are not enforcing that that data actually exists within model outputs.

Author: Happy for better wording suggestions to capture what I just mentioned though!

Reviewer 3: Maybe "must only include the `date_col`, `target_col`, and any other task ID columns"?

Author: Sure! Done in 9e707c5ad1a7263615cc22fb2c59550a43f0efb2

* `oracle-output`:

>   * `observable_unit` must only include task ID columns, the `date_col` and the `target_col` unless `target_keys` are `NULL` which implies a single global target and no `target` column.

Reviewer 3: Likewise as my previous comment on the description of the `observable_unit`

Author: See above
